File: mike/hearts/src/TableIO.hs
module TableIO where

import Data.IORef (IORef)
import qualified Data.IORef as IORef

import Cards
import GameEvent
import Table

tableIO :: [Player] -> IO (GameCommand -> IO [GameEvent])
tableIO players =
  do ref <- IORef.newIORef (tableLoopM, emptyTableState players)
     let processCommand command =          
           do (next, state) <- IORef.readIORef ref
              putStrLn ("processCommand " ++ (show command))
              let (state', events, step) = runTable (next command) state []              
              case step of
                Left cont ->
                  do putStrLn ("NeedsCommand")
                     IORef.writeIORef ref (cont, state')
                Right _result -> return ()
              return events
     return processCommand

File: hearts-scala/src/main/scala/funar/hearts/TableIO.scala
package funar.hearts

import cats.Traverse
import cats.data._
import cats.implicits._
import cats.effect.IO

object TableIO {
  import Table._
  def tableIO(players: List[Player]): GameCommand => IO[Seq[GameEvent]] = {
    var ref = (tableLoopM, emptyTableState(players))
    def processCommand(command: GameCommand): IO[Seq[GameEvent]] = IO {
      val (next, state) = ref
      val (state1, events, step) = runTable(next(command), state, Seq.empty)
      step match {
        case Left(cont) => ref = (cont, state1)
        case Right(_) => ()
      }
      events
    }
    processCommand
  }
}